cmd/query/app/server.go

```diff
@@ -69,19 +80,39 @@ func NewServer(querySvc *querysvc.QueryService,
 		return nil, fmt.Errorf("invalid gRPC server host:port: %w", err)
 	}
 	separatePorts := grpcPort != httpPort || grpcPort == "0" || httpPort == "0"
+	if !separatePorts && !allowSamePort {
+		return nil, ErrSameGRPCandHTTPPort
+	}
 
 	if (options.HTTP.TLSSetting != nil || options.GRPC.TLSSetting != nil) && !separatePorts {
 		return nil, errors.New("server with TLS enabled can not use same host ports for gRPC and HTTP.  Use dedicated HTTP and gRPC host ports instead")
 	}
 
-	grpcServer, err := createGRPCServer(querySvc, metricsQuerySvc, options, tm, telset)
-	if err != nil {
-		return nil, err
-	}
+	var grpcServer *grpc.Server
+	var httpServer *httpServer
+	if separatePorts {
+		grpcServer, err = createGRPCServer(ctx, host, querySvc, metricsQuerySvc, options, tm, telset)
```

Comment:
I think this is unnecessary bifurcation at this point. I would instead spit this function into two, `create` and `registerEndpoints(server)`. Pass `legacy=separatePorts` argument to `create`
```
func createGRPCServer(legacy bool, ... {
  if !legacy {
    return otel.Create...
  }
  // legacy code
}
```

Reply:
@yurishkuro how would we register the endpoints after creating the server?

Reply:
Same way they are currently registered in the  create() function. My point is that we're only changing how the server is created, not how endpoints are mounted into it.

Reply:
@yurishkuro otel's `ToServer` takes in a handler, which has the routes registered to it. So how would we register the routes after creating the server using `ToServer`?

Reply:
ok, so for grpc the order it (1) create server, (2) register handlers. For HTTP it's (1) create the root handler with all sub-handlers registered, and (2) create server. In both cases we can factor out the handlers creation, which are not different between v1/v2.